agent/tool-scripts/jmap

```diff
@@ -0,0 +1,168 @@
+# 3) Stop data collection
+# 4) post-process the data
+
+# Defaults
+tool=jmap
+tool_bin=/usr/bin/$tool
+group=default
+dir=""
+mode=""
+iteration="1"
+options="none"
+jcn=""
+
+function usage() {
+echo "
+USAGE:
+	$script_name: you specified an invalid option
+	The following options are available:
+	--install,			        install this perf tool
+	--start|stop|postprocess	start/stop/post-process the data collection
+	--iteration=int 		    the iteration (required)
+	--group=str		            the perftool group (not-required)
+	--dir=str			        directory to store data collection (required)
+	--pid=int		            a process ID to profile
+	--pattern			        collect jvm stats for any PID which name matches this string
+	--jvm-container-name             collect jvm stats for containerized jvm which name matches this string(container name)
+"
+}
+
+
+# Process options and arguments
+opts=$(getopt -q -o idp --longoptions "dir:,group:,iteration:,pid:,pattern:,start,stop,install,postprocess" -n "getopt.sh" -- "$@");
+if [ $? -ne 0 ]; then
+        usage
+        exit 1
+fi
+
+eval set -- "$opts";
+while true; do
+	case "$1" in
+		--install)
+		mode="install"
+		shift;
+		;;
+		--start)
+		mode="start"
+		shift;
+		;;
+		--stop)
+		mode="stop"
+		shift;
+		;;
+		--postprocess)
+		mode="postprocess"
+		shift;
+		;;
+		-d|--dir)
+		shift;
+		if [ -n "$1" ]; then
+			dir="$1"
+			shift
+		fi
+		;;
+		-g|--group)
+		shift;
+		if [ -n "$1" ]; then
+			group="$1"
+			shift
+		fi
+		;;
+		-i|--iteration)
+		shift;
+		if [ -n "$1" ]; then
+			iteration="$1"
+			shift
+		fi
+		;;
+		--pid)
+		shift;
+		if [ -n "$1" ]; then
+			jvm_pid="$1"
+			shift
+		fi
+		;;
+		--pattern)
+		shift;
+		if [ -n "$1" ]; then
+			pattern="$1"
+			shift
+		fi
+		;;
+		-j|--jvm-container-name)
+		shift;
+		if [ -n "$1" ]; then
+			jcn="$1"
+			shift
+		fi
+		;;
+		--)
+		shift;
+		break;
+		;;
+		*)
+		break;
+	esac
+done
+
+tool_dir="$dir/tools-$group"
+tool_output_dir=$tool_dir/$tool # all tools keep data in their tool specific dir
+case "$mode" in
+	install)
+	    yum install -y --nogpgcheck java
+        ;;
+	start)
+	if [ ! -z "$pattern" ]; then
+		jvm_pids=`ps aux | pgrep $pattern`
+	fi
+	jvm_pids="$jvm_pids $jvm_pid"
+	mkdir -p $tool_output_dir
+    	echo "$script_name: running $tool_cmd"
+	for jvm_pid in $jvm_pids; do
+		tool_stdout_file=$tool_output_dir/$tool-stdout.txt
+		tool_stderr_file=$tool_output_dir/$tool-stderr.txt
+		tool_pid_file=$pbench_tmp/$group.$iteration.$tool-$jvm_pid.pid
+		tool_cmd_file="$tool_output_dir/$tool-$jvm_pid.cmd"
+		if [ ! -z "$jcn" ]; then
+			x
+                        heap_used=$(docker exec -ti `docker ps --filter name=$jcn --format "{{.Names}}"` $tool_bin -histo $jpid |grep Total |awk '{print $3/1024/1024}')
+                else
+                        heap_used=$($tool_bin -histo $jvm_pid |grep Total |awk '{print $3/1024/1024}')
+                fi
+                echo "$(date "+%s"),$jvm_pid,$heap_used" >> "$tool_stdout_file"
```

Comment:
So this tool collects the jmap output for a set of JVM pids and then exits?  So it basically collects the output once at the beginning of the start tools, and then stops running?

Reply:
hmmm, I assume start means run as long the agent is running.
on the other hand I can't see any loop that runs it forever the agent is running.

Reply:
From the code, it seems "starting a tool" is performed before a benchmark script is run, and "stopping a tool" is performed after a benchmark script completes.  Those two events give the tool script a chance to do whatever that requires in order to execute.

If this jmap tool needs to run periodically to collect heap metrics, then we'll need a background process to do that. Then the "start tool" will create that process and the "stop tool" will kill that process.